timemachine/cpp/src/verlet_integrator.cu

```diff
@@ -0,0 +1,119 @@
+}
+
+VelocityVerletIntegrator::~VelocityVerletIntegrator() {
+    gpuErrchk(cudaFree(d_cbs_));
+    gpuErrchk(cudaFree(d_du_dx_));
+}
+
+void VelocityVerletIntegrator::step_fwd(
+    std::vector<BoundPotential *> &bps,
+    double lamb,
+    double *d_x_t,
+    double *d_v_t,
+    double *d_box_t,
+    unsigned long long *d_du_dl,
+    cudaStream_t stream) {
+
+    gpuErrchk(cudaMemsetAsync(d_du_dx_, 0, N_ * 3 * sizeof(*d_du_dx_), stream));
+
+    const int D = 3;
+    size_t tpb = warp_size;
+    size_t n_blocks = ceil_divide(N_, tpb);
+    dim3 dimGrid_dx(n_blocks, D);
+    for (int i = 0; i < bps.size(); i++) {
+        bps[i]->execute_device(
+            N_, d_x_t, d_box_t, lamb, d_du_dx_, nullptr, d_du_dl ? d_du_dl : nullptr, nullptr, stream);
+    }
+    update_forward_velocity_verlet<double><<<dimGrid_dx, tpb, 0, stream>>>(N_, D, d_cbs_, d_x_t, d_v_t, d_du_dx_, dt_);
+    gpuErrchk(cudaPeekAtLastError());
+}
+
+void VelocityVerletIntegrator::initialize(
+    std::vector<BoundPotential *> &bps,
+    double lamb,
+    double *d_x_t,
+    double *d_v_t,
+    double *d_box_t,
+    cudaStream_t stream) {
+
+    if (initialized_) {
+        // No op if this is initialized, will just mess up the velos
+        return;
+    }
+    gpuErrchk(cudaMemsetAsync(d_du_dx_, 0, N_ * 3 * sizeof(*d_du_dx_), stream));
+
+    const int D = 3;
+    size_t tpb = warp_size;
+    size_t n_blocks = ceil_divide(N_, tpb);
+    dim3 dimGrid_dx(n_blocks, D);
+
+    for (int i = 0; i < bps.size(); i++) {
+        bps[i]->execute_device(
+            N_,
+            d_x_t,
+            d_box_t,
+            lamb,
+            d_du_dx_, // we only need the forces
+            nullptr,
+            nullptr,
+            nullptr,
+            stream);
+    }
+    half_step_velocity_verlet<double, true><<<dimGrid_dx, tpb, 0, stream>>>(N_, D, d_cbs_, d_x_t, d_v_t, d_du_dx_, dt_);
+    gpuErrchk(cudaPeekAtLastError());
+    initialized_ = true;
+};
+
+void VelocityVerletIntegrator::finalize(
+    std::vector<BoundPotential *> &bps,
+    double lamb,
+    double *d_x_t,
+    double *d_v_t,
+    double *d_box_t,
+    cudaStream_t stream) {
+
+    if (!initialized_) {
+        // No op if this isn't initialized, will just mess up the velos
+        return;
+    }
+    gpuErrchk(cudaMemsetAsync(d_du_dx_, 0, N_ * 3 * sizeof(*d_du_dx_), stream));
+
+    const int D = 3;
+    size_t tpb = warp_size;
+    size_t n_blocks = ceil_divide(N_, tpb);
+    dim3 dimGrid_dx(n_blocks, D);
+
+    for (int i = 0; i < bps.size(); i++) {
+        bps[i]->execute_device(
+            N_,
+            d_x_t,
+            d_box_t,
+            lamb,
+            d_du_dx_, // we only need the forces
+            nullptr,
+            nullptr,
+            nullptr,
+            stream);
+    }
+    half_step_velocity_verlet<double, false>
+        <<<dimGrid_dx, tpb, 0, stream>>>(N_, D, d_cbs_, d_x_t, d_v_t, d_du_dx_, dt_);
+    gpuErrchk(cudaPeekAtLastError());
+    initialized_ = false;
```

Comment:
Q: is it legal to call `finalize` immediately followed by `initialize` to resume a trajectory? Python pseudocode:
```python
integ = VelocityVerletIntegrator()
integ.initialize()
for _ in range(100):
    integ.step_fwd()
integ.finalize()
integ.initialize()
for _ in range(100):
    integ.step_fwd()
integ.finalize()
```
and is that the same as running 200 steps without finalize/initialize in between?

Reply:
Should be perfectly fine.

The trajectory is defined by a for loop, where the loop body is three steps

____
V(dt/2)
R(dt)
V(dt/2)
____
V(dt/2)
R(dt)
V(dt/2)
____
V(dt/2)
R(dt)
V(dt/2)
____
...
____
V(dt/2)
R(dt)
V(dt/2)
____

However, for implementation reasons it is more convenient if the main loop body only requires the forces computed the previous step. We can accomplish this by manually applying a "loop shifting" optimization:
____
V(dt/2)
R(dt)
____
V(dt/2)
V(dt/2)
R(dt)
____
V(dt/2)
V(dt/2)
R(dt)
____
...
____
V(dt/2)
V(dt/2)
R(dt)
____
V(dt/2)
____

This just causes a small complication https://github.com/proteneer/timemachine/blob/53495d130c70d010ce1e04c96d72d90fa4441687/timemachine/integrator.py#L158-L191 where the state at any point between iterations of the main for-loop will have the positions and velocities staggered by half a timestep. Regardless, "initialize; main loop; finalize;" should implement the exact same sequence of updates as the defined trajectory, just grouping the updates differently for implementation convenience.